<a href="https://colab.research.google.com/github/virf96/Chat-Bot/blob/master/Datos_Faltantes_MetodoSustitucionDeteccionAutomatica_Sklearn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Selección automática de la mejor técnica para sustitución con Sklearn

En este notebook vamos a hacer búsqueda exhaustiva de hiper-parámetros o búsqueda de cuadricula por su nombre en inglés 'grid search' sobre los métodos de sustitución disponibles en Scikit-learn para determinar cuáles técnicas de imputación funcionan mejor para este conjunto de datos y el modelo de aprendizaje de máquina que seleccionemos

También vamos a entrenar un modelo simple de machine learning como parte de un flujo de trabajo o 'pipeline'.


## En este demo:

Vamos a aprender a **seleccionar automáticamente la mejor técnica de sustitución utilizando Scikit-learn** y los datos Ames House Price.

In [1]:
import pandas as pd
import numpy as np

# importar clases para sustitución
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

# importar extra clases para modelamiento
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.linear_model import Lasso
from sklearn.model_selection import train_test_split, GridSearchCV

np.random.seed(0)

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
# carguemos los datos 
data = pd.read_csv('/content/drive/My Drive/datasets/houseprice.csv')
print(data.shape)
data.head()

(1460, 81)


Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [6]:
# encontrar las variables categóricas
# aquellas de tipo 'Object'  en los datos

features_categorical = [c for c in data.columns if data[c].dtypes=='O']

# encontrar las variables numéricas
# aquellas que son diferentes de  'Object' excluyendo también el target SalePrice

features_numerical = [c for c in data.columns if data[c].dtypes!='O' and c !='SalePrice']

In [7]:
# revisemos las variables categóricas

data[features_categorical].head()

MSZoning Street Alley LotShape  ... Fence MiscFeature SaleType SaleCondition
0       RL   Pave   NaN      Reg  ...   NaN         NaN       WD        Normal
1       RL   Pave   NaN      Reg  ...   NaN         NaN       WD        Normal
2       RL   Pave   NaN      IR1  ...   NaN         NaN       WD        Normal
3       RL   Pave   NaN      IR1  ...   NaN         NaN       WD       Abnorml
4       RL   Pave   NaN      IR1  ...   NaN         NaN       WD        Normal

[5 rows x 43 columns]

In [8]:
# revisemos las variables numéricas

data[features_numerical].head()

Id  MSSubClass  LotFrontage  LotArea  ...  PoolArea  MiscVal  MoSold  YrSold
0   1          60         65.0     8450  ...         0        0       2    2008
1   2          20         80.0     9600  ...         0        0       5    2007
2   3          60         68.0    11250  ...         0        0       9    2008
3   4          70         60.0     9550  ...         0        0       2    2006
4   5          60         84.0    14260  ...         0        0      12    2008

[5 rows x 37 columns]

In [9]:
# separar datos en segmentos entrenamiento y prueba
#Esto dado que el imputador solo debe de aprender del set de entrenamiento
X_train, X_test, y_train, y_test = train_test_split(
    data.drop('SalePrice', axis=1),  # removemos el target del set de datos
    data['SalePrice'],  # el target
    test_size=0.3,  
    random_state=0)  

X_train.shape, X_test.shape

((1022, 80), (438, 80))

In [10]:
# Creemos el flujo de trabajo o tubería de pre-procesamiento (preprocessing pipelines) para las variables categóricas y numéricas

# adaptado de  Scikit-learn: código disponible aquí bajo la licencia BSD3 :
# https://scikit-learn.org/stable/auto_examples/compose/plot_column_transformer_mixed_types.html

numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())])#Utilizamos StandarScaler() dado que las variables deben de tener la misma magnitud

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))])#Utilizamos OneHotEncoder para codificar las categorías en números

preprocessor = ColumnTransformer(
    transformers=[
        ('numerical', numeric_transformer, features_numerical),
        ('categorical', categorical_transformer, features_categorical)])

# Para iniciar el pipeline es necesario pasar un argumento a los transformadores
# Estos serán modificados durante la búsqueda de cuadricula gridsearch a continuación.

In [11]:
# Anadir un clasificador al preprocessing pipeline.
# ahora tenemos un flujo de trabajo (pipeline) completo para predicción.

clf = Pipeline(steps=[('preprocessor', preprocessor),
                      ('classifier', Lasso(max_iter=2000))])

In [12]:
# creemos la cuadrícula con todos los parámetros que queremos probar

param_grid = {
    'preprocessor__numerical__imputer__strategy': ['mean', 'median'],
    'preprocessor__categorical__imputer__strategy': ['most_frequent', 'constant'],
    'classifier__alpha': [10, 100, 200],
}

grid_search = GridSearchCV(clf, param_grid, cv=5, iid=False, n_jobs=-1, scoring='r2')

# cv=3 es la validación cruzada (cross-validation)
# no_jobs =-1 usar todas las CPUs disponibles  
# scoring='r2' indica que se evalué usando r cuadrado

# para más detalles sobre la cuadrícula de parámetros visitar:
#https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html


Cuando estamos fijando los parámetros de la búsqueda exhaustiva, así es que indicamos los parámetros:

- preprocessor__numerical__imputer__strategy': ['mean', 'median'],

la línea de código anterior indica que queremos probar el 'imputer' con la media y la mediana en el paso de sustitución del procesador numérico 

- preprocessor__categorical__imputer__strategy': ['most_frequent', 'constant']

la línea de código anterior indica que queremos probar la sustitución por categórica más frecuente o un valor constante en el paso del sustitución del procesador para variables calóricas

- classifier__alpha': [0.1, 1.0, 0.5]

la línea de código anterior indica que queremos probar esos 3 valores del parámetro alfa para Lasso. Nota que Lasso es el paso de modelamiento o 'regresor' de nuestra tubería o pipeline


In [13]:
# ahora vamos a entrenar sobre todas las posibles combinaciones
# de los parámetros anteriores
grid_search.fit(X_train, y_train)

# imprimimos el mejor score o puntaje sobre el segmento de prueba
print(("mejor regresión linear de la busqueda exahustiva: %.3f"
       % grid_search.score(X_train, y_train)))

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_search.py:823: FutureWarning: The parameter 'iid' is deprecated in 0.22 and will be removed in 0.24.
  "removed in 0.24.", FutureWarning


mejor regresión linear de la busqueda exahustiva: 0.933


In [14]:
# podemos imprimir los mejores parametros estimados asi

grid_search.best_estimator_

Pipeline(memory=None,
         steps=[('preprocessor',
                 ColumnTransformer(n_jobs=None, remainder='drop',
                                   sparse_threshold=0.3,
                                   transformer_weights=None,
                                   transformers=[('numerical',
                                                  Pipeline(memory=None,
                                                           steps=[('imputer',
                                                                   SimpleImputer(add_indicator=False,
                                                                                 copy=True,
                                                                                 fill_value=None,
                                                                                 missing_values=nan,
                                                                                 strategy='median',
                                                       

In [15]:
# los mejores parámetros ajustados asi
grid_search.best_params_

{'classifier__alpha': 100,
 'preprocessor__categorical__imputer__strategy': 'constant',
 'preprocessor__numerical__imputer__strategy': 'median'}

In [16]:
# podemos ver todas las combinaciones evaluadas durante el gridsearch
grid_search.cv_results_['params']

[{'classifier__alpha': 10,
  'preprocessor__categorical__imputer__strategy': 'most_frequent',
  'preprocessor__numerical__imputer__strategy': 'mean'},
 {'classifier__alpha': 10,
  'preprocessor__categorical__imputer__strategy': 'most_frequent',
  'preprocessor__numerical__imputer__strategy': 'median'},
 {'classifier__alpha': 10,
  'preprocessor__categorical__imputer__strategy': 'constant',
  'preprocessor__numerical__imputer__strategy': 'mean'},
 {'classifier__alpha': 10,
  'preprocessor__categorical__imputer__strategy': 'constant',
  'preprocessor__numerical__imputer__strategy': 'median'},
 {'classifier__alpha': 100,
  'preprocessor__categorical__imputer__strategy': 'most_frequent',
  'preprocessor__numerical__imputer__strategy': 'mean'},
 {'classifier__alpha': 100,
  'preprocessor__categorical__imputer__strategy': 'most_frequent',
  'preprocessor__numerical__imputer__strategy': 'median'},
 {'classifier__alpha': 100,
  'preprocessor__categorical__imputer__strategy': 'constant',
  'pre

In [17]:
# y ahora los puntajes para cada una de las combinaciones anteriores
grid_search.cv_results_['mean_test_score']

array([0.84746254, 0.84739594, 0.84814964, 0.8481309 , 0.86624908,
       0.86621021, 0.86646886, 0.86651035, 0.86552764, 0.8654755 ,
       0.86525292, 0.86523714])

In [18]:
# y finalmente miremos el desempeño en el segmento de prueba
print(("best linear regression from grid search: %.3f"
       % grid_search.score(X_test, y_test)))

best linear regression from grid search: 0.738


Este modelo sobre-ajusta el segmento de enteramiento; mirando el valor de r2 de 0.93 que obtuvimos en el segmento de entrenamiento vs. 0.738 para el segmento de prueba.

Trataremos de reducir este sobre-ajuste en otra demo.